src/operations/client_bulk_write/common.ts

```diff
@@ -0,0 +1,148 @@
+  filter: Filter<TSchema>;
+  /** Specifies a collation. */
+  collation?: CollationOptions;
+  /** The index to use. If specified, then the query system will only consider plans using the hinted index. */
+  hint?: Hint;
+}
+
+/** @public */
+export interface ClientReplaceOneModel<TSchema extends Document = Document>
+  extends ClientWriteModel {
+  name: 'replaceOne';
+  /**
+   * The filter used to determine if a document should be replaced.
+   * For a replaceOne operation, the first match is replaced.
+   */
+  filter: Filter<TSchema>;
+  /** The document with which to replace the matched document. */
+  replacement: WithoutId<TSchema>;
+  /** Specifies a collation. */
+  collation?: CollationOptions;
+  /** The index to use. If specified, then the query system will only consider plans using the hinted index. */
+  hint?: Hint;
+  /** When true, creates a new document if no document matches the query. */
+  upsert?: boolean;
+}
+
+/** @public */
+export interface ClientUpdateOneModel<TSchema extends Document = Document>
+  extends ClientWriteModel {
+  name: 'updateOne';
+  /**
+   * The filter used to determine if a document should be updated.
+   * For an updateOne operation, the first match is updated.
+   */
+  filter: Filter<TSchema>;
+  /**
+   * The modifications to apply. The value can be either:
+   * UpdateFilter<TSchema> - A document that contains update operator expressions,
+   * Document[] - an aggregation pipeline.
+   */
+  update: UpdateFilter<TSchema> | Document[];
+  /** A set of filters specifying to which array elements an update should apply. */
+  arrayFilters?: Document[];
+  /** Specifies a collation. */
+  collation?: CollationOptions;
+  /** The index to use. If specified, then the query system will only consider plans using the hinted index. */
+  hint?: Hint;
+  /** When true, creates a new document if no document matches the query. */
+  upsert?: boolean;
+}
+
+/** @public */
+export interface ClientUpdateManyModel<TSchema extends Document = Document>
+  extends ClientWriteModel {
+  name: 'updateMany';
+  /**
+   * The filter used to determine if a document should be updated.
+   * For an updateMany operation, all matches are updated.
+   */
+  filter: Filter<TSchema>;
+  /**
+   * The modifications to apply. The value can be either:
+   * UpdateFilter<TSchema> - A document that contains update operator expressions,
+   * Document[] - an aggregation pipeline.
+   */
+  update: UpdateFilter<TSchema> | Document[];
+  /** A set of filters specifying to which array elements an update should apply. */
+  arrayFilters?: Document[];
+  /** Specifies a collation. */
+  collation?: CollationOptions;
+  /** The index to use. If specified, then the query system will only consider plans using the hinted index. */
+  hint?: Hint;
+  /** When true, creates a new document if no document matches the query. */
+  upsert?: boolean;
+}
+
+/** @public */
+export type AnyClientBulkWriteModel<TSchema extends Document = Document> =
```

Comment:
Can we add documentation for what is `TSchema` means for this type? 

IIUC Each one of these models has a `namespace` property that can be different, that probably means there's a mixture of schemas, how does the use case for a `ClientUpdateOneModel<Books>` and `ClientDeleteOneModel<Authors>` work?

Reply:
Docs added.

Reply:
```ts
type Book = { _id: UUID };
type Author = { _id: ObjectId };

const array: AnyClientBulkWriteModel<Book | Author> = [
  { namespace: 'db.books', name: 'insertOne', document: { _id: new ObjectId() } },
  { namespace: 'db.authors', name: 'updateOne', filter: {}, update: { $set: { _id: new UUID() } } }
];
```

(ignoring the fact that $set doesn't make sense for _id)

If I just union my schemas together then I don't get safety when there is overlap, so I don't see how in practice this API has a good DX for the typescript user. 

But we did say:

> TSchema can be different on each of the individual models must always match the appropriate namespace that it defines provided to each of the models. 

Is the intention to cast or declare separate variables? Would it be possible to "lookup" the schema the way we do event listeners if the namespace string is a literal type?

Reply:
I've gone ahead and done that. Can maybe revisit options in the next ticket for the full happy path.

Reply:
I am okay with revisiting, can we actually file a ticket specifically for investigating the TS DX of the API, we can finalize the functionality of the API with weak types and consider possibly a type intended to be used as an annotation to opt into an autocomplete experience (like `StrictFilter`)